### DISCUSSION TITLE
ERROR in running METPlus GridStat

### DISCUSSION THREAD

**Discussion by wuxx66** [dtcenter/METplus#2650] [Configuration] [answered]
I am trying to verify hourly 2m temperature against analysis data using GridStat. I got an error
ERROR  :
ERROR  : NcCfFile::getData(NcVar *, const LongArray &, DataPlane &) const -> needed 3 arguments for variable data, got 2
ERROR  :

I have put the config files, log file, input files and mask file in directory wu_data

Could you help me have a look what is wrong?

Thank you so much

Xiaoxi

**Comment by jprestop** [comment]
Hi Xiaoxi (@[wuxx66](https://github.com/wuxx66)).  Thank you for describing the problem and for uploading your files to the wu_data subdirectory.  

The problem is arising when reading the poly mask file:
```
mask = {poly = ["/g/data/wa46/user/wuxx/science_test/GFE_hourly_T/mask_files/gfe_mask.nc"];}
```

That file is formatted as a CF-compliant NetCDF file which will only work with some additional information. Here’s a description of the [mask.poly config option](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html#mask)  from the User’s Guide:
```
The "poly" entry contains a comma-separated list of files that define verification masking regions. These masking regions may be specified in two ways: in an ASCII file containing lat/lon points defining the mask polygon, or using a gridded data file such as the NetCDF output of the Gen-Vx-Mask tool.
```
We need to improve this documentation to add more details and examples.  If that file were the NetCDF output from MET’s gen_vx_mask tool, it would work fine without any changes.

There are two options for how to proceed in resolving the error:

- Use the NetCDF output of MET’s  gen_vx_mask tool.
- Use the existing gridded data file followed by a configuration string to read data from the file and a threshold to convert that data into a 0/1 mask. 

For Option 1:
Gen-Vx-Mask always writes output with “(lat, lon)” dimensions which is why MET looks for those. The good news is that Gen-Vx-Mask can read CF-compliant NetCDF files. So you can run Gen-Vx-Mask to produce a masking grid that Grid-Stat will be happy with.

Running Gen-Vx-Mask would look like this:
```
gen_vx_mask gfe_mask.nc gfe_mask.nc gfe_mask_met.nc -type data -mask_field 'name="data"; level="(0,*,*)";' -thresh gt-30
```
And then you’d directly include that  output file in your Grid-Stat configuration:
```
poly = ["/d1/projects/METplus/discussions/2650/gfe_mask_met.nc”];
```
This is the resulting image produced when using `ncview` to view the resulting output gfe_mask_met.nc file:
<img width="748" alt="gfe_mask_met" src="https://github.com/user-attachments/assets/dc589c0a-db16-4f3a-a507-122dd1b6de09">

Here are some of the Grid-Stat log messages which show the number of points in the FULL domain vs the reduced number in the masking region:
```
DEBUG 2: Processing data(20230302_150000,*,*) versus data(20230302_150000,*,*), for smoothing method NEAREST(1), over region FULL, using 544260 matched pairs.
DEBUG 2: Computing Scalar Partial Sums and Continuous Statistics.
DEBUG 2: Processing data(20230302_150000,*,*) versus data(20230302_150000,*,*), for smoothing method NEAREST(1), over region data_0,*,*!=-30, using 300890 matched pairs.
DEBUG 2: Computing Scalar Partial Sums and Continuous Statistics.
DEBUG 2: Processing data(20230302_150000,*,*) versus data(20230302_150000,*,*), for smoothing method NEAREST(1), over region data_mask, using 300890 matched pairs.
DEBUG 2: Computing Scalar Partial Sums and Continuous Statistics.
```

For Option 2:

You can actually list that data file directly in the Grid-Stat configuration file, tell it what data to read, and then append the threshold to be applied. The formatting of the string can be a little finicky, however, but it would look like this:

```
poly = ["/d1/projects/METplus/discussions/2650/gfe_mask.nc {name=\"data\"; level=\"(0,*,*)\";} !=-30"];
```
Please note that “!=-30” represents the threshold to be applied.  Looking at the data in the gfe_mask.nc file, it appeared as if values of -30 represented values that wouldn’t be in the mask, so we are excluding those here.  

Since the mask is static, meaning it doesn’t change from time-step to time-step, you can just run Gen-Vx-Mask ahead of time once and save the output to use in your grid_stat runs.

Using one of these options should resolve the error you are receiving. Please let us know if you have any further questions.

**Reply by jprestop** [reply]
@wuxx66 As mentioned in my previous comment, we needed to improve our documentation with more details and examples.  The documentation has been updated and you can find it [here](https://met.readthedocs.io/en/develop/Users_Guide/config_options.html#mask).

**Reply by wuxx66** [reply]
Hi Julie, thank you for identifying and explaining the issue in my case. I have recreated the mask file and the task is running fine now. Also thank for updating the documentation.

**Reply by jprestop** [reply, marked as answer]
@wuxx66 Great!  Thank you for letting us know.  If you don't have any further questions at this time, please go ahead and click the "Mark as answer" button. Then, I will LOCK this conversation to prevent future posts. Our team has decided that once discussions have been answered, and the answer has been confirmed, we’ll lock them. We want to encourage users to ask new questions in new discussions rather than posting to old ones. Hopefully that’ll make the questions/answers easier for other users to follow. So if/when more issues/questions arise, please feel free to start a new discussion.
